src/lib/db/role-store.ts

```diff
@@ -0,0 +1,196 @@
+            [id],
+        );
+        const { present } = result.rows[0];
+        return present;
+    }
+
+    async nameInUse(name: string, existingId?: number): Promise<boolean> {
+        let query = this.db(T.ROLES).where({ name }).returning('id');
+        if (existingId) {
+            query = query.andWhereNot({ id: existingId });
+        }
+        const result = await query;
+        return result.length > 0;
+    }
+
+    async roleExists(name: string): Promise<boolean> {
+        const result = await this.db.raw(
+            `SELECT EXISTS (SELECT 1 FROM ${T.ROLES} WHERE name = ?) AS present`,
+            [name],
+        );
+        const { present } = result.rows[0];
+        return present;
+    }
+
+    async deleteAll(): Promise<void> {
+        return this.db(T.ROLES).del();
+    }
+
+    mapRow(row: IRoleRow): ICustomRole {
+        if (!row) {
+            throw new NotFoundError('No row');
+        }
+
+        return {
+            id: row.id,
+            name: row.name,
+            description: row.description,
+            type: row.type,
+        };
+    }
+
+    async getRoles(): Promise<IRole[]> {
+        return this.db
+            .select(['id', 'name', 'type', 'description'])
+            .from<IRole>(T.ROLES);
+    }
+
+    async getRoleWithId(id: number): Promise<IRole> {
+        return this.db
+            .select(['id', 'name', 'type', 'description'])
+            .where('id', id)
+            .first()
+            .from<IRole>(T.ROLES);
+    }
+
+    async getProjectRoles(): Promise<IRole[]> {
+        return this.db
+            .select(['id', 'name', 'type', 'description'])
+            .from<IRole>(T.ROLES)
+            .where('type', 'custom')
+            .orWhere('type', 'project');
+    }
+
+    async getRolesForProject(projectId: string): Promise<IRole[]> {
+        return this.db
+            .select(['r.id', 'r.name', 'r.type', 'ru.project', 'r.description'])
+            .from<IRole>(`${T.ROLE_USER} as ru`)
+            .innerJoin(`${T.ROLES} as r`, 'ru.role_id', 'r.id')
+            .where('project', projectId);
+    }
+
+    async getRootRoles(): Promise<IRole[]> {
+        return this.db
+            .select(['id', 'name', 'type', 'description'])
+            .from<IRole>(T.ROLES)
+            .where('type', 'root');
+    }
+
+    async removeRolesForProject(projectId: string): Promise<void> {
+        return this.db(T.ROLE_USER)
+            .where({
+                project: projectId,
+            })
+            .delete();
+    }
+
+    async getRootRoleForAllUsers(): Promise<IUserRole[]> {
+        const rows = await this.db
+            .select('id', 'user_id')
+            .distinctOn('user_id')
+            .from(`${T.ROLES} AS r`)
+            .leftJoin(`${T.ROLE_USER} AS ru`, 'r.id', 'ru.role_id')
+            .where('r.type', '=', 'root');
+
+        return rows.map((row) => ({
+            roleId: +row.id,
```

Comment:
Clever way of casting to a number, but Number(row.id) would perhaps be more readable?

Reply:
I had to do a double on this one because I don't remember writing it and I wasn't actually sure what the '+; did, turns out I just moved it to a different file!

So I learned a new thing today - that you can use + to cast a string to an int.

And I also totally agree and have swapped this one out for Number(row.id)